sklearn_numba_dpex/common/kernels.py

```diff
@@ -223,43 +232,90 @@ def make_sum_reduction_2d_axis1_kernel(
-    then the expected shape of input tensors is `(size0, size1)`, and the reduction
-    operation is equivalent to input.sum(axis=1). Depending on the size of the sum
-    axis, it might be worth tuning `work_group_size` for better performance.
+    .. [3] https://shreeraman-ak.medium.com/parallel-reduction-with-cuda-d0ae10c1ae2c
     """
-    if size1 is not None:
-        n_rows = size0
+    if is_1d := not size1:
+        axis = 1
+        size1 = size0
+        size0 = 1
+
+    # NB: the shape of the work group is different in each of those two cases. Summing
+    # efficiently requires to adapt to very different IO patterns depending on the sum
+    # axis, which motivates the need for different kernels with different work group
+    # sizes for each cases. As a consequence, the shape of the intermediate results
+    # in the main driver loop, and the `global_size` (total number of work items fired
+    # per call) for each kernel call, are also differents, and are variabilized with
+    # the lambda functions `get_result_shape1 and `get_global_size`.
+    if axis == 1:
+        work_group_size, kernels = _prepare_sum_reduction_2d_axis1(
+            size0, work_group_size, fused_elementwise_func, dtype, device
+        )
         sum_axis_size = size1
-    else:
-        n_rows = 1
-        sum_axis_size = size0
+        get_result_shape = lambda result_sum_axis_size: (size0, result_sum_axis_size)
+        get_global_size = (
+            lambda result_sum_axis_size: result_sum_axis_size * work_group_size * size0
+        )
 
-    # fused_unary_func is applied elementwise during the first pass on data, in the
-    # first kernel execution only.
-    work_group_size, fused_func_kernel = _make_partial_sum_reduction_2d_axis1_kernel(
-        n_rows, work_group_size, fused_unary_func, dtype, device
-    )
-    # subsequent kernel calls only sum the data.
-    _, nofunc_kernel = _make_partial_sum_reduction_2d_axis1_kernel(
-        n_rows, work_group_size, fused_unary_func=None, dtype=dtype, device=device
-    )
+    else:  # axis == 0
+        work_group_size, kernels = _prepare_sum_reduction_2d_axis0(
+            size1,
+            work_group_size,
+            sub_group_size,
+            fused_elementwise_func,
+            dtype,
+            device,
+        )
+        sum_axis_size = size0
+        get_result_shape = lambda result_sum_axis_size: (result_sum_axis_size, size1)
+        get_global_size = (
+            lambda result_sum_axis_size: result_sum_axis_size
+            * work_group_size
+            * math.ceil(size1 / sub_group_size)
+        )
 
-    check_power_of_2(work_group_size)
+    # `fused_elementwise_func` is applied elementwise during the first pass on
+    # data, in the first kernel execution only, using `fused_func_kernel`. Subsequent
+    # kernel calls only sum the data, using `nofunc_kernel`.
+    (fused_func_kernel, nofunc_kernel), reduction_block_size = kernels
```

Comment:
I find this unpacking unconventional and here my proposal to improve it (3/3).
```suggestion
    (fused_func_kernel, nofunc_kernel)  = kernels
```